src/gpio.rs

```diff
@@ -293,1040 +377,841 @@ macro_rules! gpio {
-                    }
-                }
+impl<MODE, CR, const P: char, const N: u8> InputPin for Pin<Input<MODE>, CR, P, N> {
+    type Error = Infallible;
+    #[inline]
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        Ok(self.is_high())
+    }
 
-                // embedded_hal impls
+    #[inline]
+    fn is_low(&self) -> Result<bool, Self::Error> {
+        Ok(self.is_low())
+    }
+}
 
-                impl<MODE> OutputPin for $PXi<Output<MODE>> {
-                    type Error = Infallible;
-                    #[inline]
-                    fn set_high(&mut self) -> Result<(), Self::Error> {
-                        // NOTE(unsafe) atomic write to a stateless register
-                        Ok(self.set_state(State::High))
-                    }
+impl<CR, const P: char, const N: u8> Pin<Output<OpenDrain>, CR, P, N> {
+    #[inline]
+    pub fn is_high(&self) -> bool {
+        !self._is_low()
+    }
+    #[inline]
+    pub fn is_low(&self) -> bool {
+        self._is_low()
+    }
+}
 
-                    #[inline]
-                    fn set_low(&mut self) -> Result<(), Self::Error> {
-                        // NOTE(unsafe) atomic write to a stateless register
-                        Ok(self.set_state(State::Low))
-                    }
-                }
+impl<CR, const P: char, const N: u8> InputPin for Pin<Output<OpenDrain>, CR, P, N> {
+    type Error = Infallible;
+    #[inline]
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        Ok(self.is_high())
+    }
 
-                impl<MODE> StatefulOutputPin for $PXi<Output<MODE>> {
-                    #[inline]
-                    fn is_set_high(&self) -> Result<bool, Self::Error> {
-                        self.is_set_low().map(|b| !b)
-                    }
+    #[inline]
+    fn is_low(&self) -> Result<bool, Self::Error> {
+        Ok(self.is_low())
+    }
+}
 
-                    #[inline]
-                    fn is_set_low(&self) -> Result<bool, Self::Error> {
-                        Ok(self._is_set_low())
-                    }
-                }
+/// Opaque CR register
+pub struct Cr<CR, const P: char> {
+    _cr: PhantomData<CR>,
+}
 
-                impl<MODE> OutputSpeed<$CR> for $PXi<Output<MODE>> {
-                    fn set_speed(&mut self, cr: &mut $CR, speed: IOPinSpeed){
-                        const OFFSET: u32 = (4 * $i) % 32;
+macro_rules! cr {
+    ($CR:ident, $cr:ident) => {
+        #[doc(hidden)]
+        pub struct $CR {
+            _0: (),
+        }
 
-                        cr.cr().modify(|r, w| unsafe {
-                            w.bits((r.bits() & !(0b11 << OFFSET)) | ((speed as u32) << OFFSET))
-                        });
-                    }
-                }
+        impl<const P: char> Cr<$CR, P> {
+            // NOTE(allow) we get a warning on GPIOC because it only has 3 high pins
+            #[allow(dead_code)]
+            pub(crate) fn cr(&mut self) -> &pac::gpioa::$CR {
```

Comment:
Maybe I'm misunderstanding things, but should this really be `gpioa`, seems like this is more generic?

Reply:
Sorry. Didn't notice.